### DISCUSSION TITLE
Inserir dados com Entity em tabelas n para n

### DISCUSSION THREAD

**Discussion by RochaRaphael** [balta-io/forum#110] [Tira Dúvidas] [answered]
Olá, estou fazendo um programa com uma tabela cadastrando nome da pessoa e dados de contato e a outra a skill que a pessoa tem, sendo assim, uma pessoa pode ter várias skills diferentes e uma skill pode ter várias pessoas. 
   
   Ao tentar inserir os dados da pessoa eu não consigo adicionar a skill dela diretamente, apenas de forma isolada. primeiro adiciono na tabela skill e depois adiciono na tabela contact.  Isso faz com que não faça a relação entre os dois.

Gostaria de entender qual é a melhor forma de inserir um dado nesse caso da relação de n para n.

 No curso não consegui entender como fazer na prática essa inserção no caso de muitos para muitos

Esse é um exemplo de uma das minhas várias tentativas:

     using (var context = new NatworkingListDataContext())
            {
                var contato1 = new Contact
                {
                    Name = "Raquel",
                    Email = "raquel@outlook.com",
                    Age = 19,
                    Cellphone = "123453339",
                    Skills = new Skill{
                        SkillName = "Dotnet",
                        YourSkillScore = 8
                    }
                };
            }

Aqui está o link do projeto com o FluntMap: https://github.com/RochaRaphael/NetworkingList

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @RochaRaphael, como vai? No caso, era para o EF fazer isto automaticamente... como está seu mapeamento? Cola o código aqui por favor.

**Reply by RochaRaphael** [reply]
Segue abaixo o mapeamento.  
OBS: No models Ambos estão ligados com o IList exemplo, o contact tem a propriedade 

      public IList<Skill> Skills { get; set; } 

e vice-versa 

### Mapping
 ContactMap

        using Microsoft.EntityFrameworkCore;
        using NetworkingList.Models;

       namespace NetworkingList.Data.Mappings
      {
      public class ContactMap : IEntityTypeConfiguration<Contact>
      {
               public void Configure(Microsoft.EntityFrameworkCore.Metadata.Builders.EntityTypeBuilder<Contact> builder)
        {
                builder.ToTable("Contact");

            //Primary Key
            builder.HasKey(x => x.Id);

            builder.Property(x => x.Id)
                .ValueGeneratedOnAdd()
                .UseIdentityColumn();

            builder.Property(x => x.Name)
                .IsRequired()
                .HasColumnName("Name")
                .HasColumnType("NVARCHAR")
                .HasMaxLength(80);

            builder.Property(x => x.Email)
                .HasColumnName("Email")
                .HasColumnType("NVARCHAR")
                .HasMaxLength(200);

            builder.Property(x => x.Cellphone)
                .HasColumnName("Cellphone")
                .HasColumnType("VARCHAR")
                .HasMaxLength(14);

            builder.Property(x => x.Age)
                .HasColumnName("Age")
                .HasColumnType("INTEGER")
                .HasMaxLength(2);

            builder
                .HasMany(x => x.Skills)
                .WithMany(x => x.Contacts)
                .UsingEntity<Dictionary<string, object>>
                (
                    "ContactSkills",
                    contact => contact
                        .HasOne<Skill>()
                        .WithMany()
                        .HasForeignKey("ContactsId")
                        .HasConstraintName("FK_ContractSkills_ContractId"),
                    skill => skill
                        .HasOne<Contact>()
                        .WithMany()
                        .HasForeignKey("SkillId")
                        .HasConstraintName("FK_ContractSkills_SkillId")
                );
            }}}
   

**SkillMap**

     using Microsoft.EntityFrameworkCore;
     using Microsoft.EntityFrameworkCore.Metadata.Builders;
     using NetworkingList.Models;

    namespace NetworkingList.Data.Mappings
    {
         public class SkillMap : IEntityTypeConfiguration<Skill>
         {
              public void Configure(EntityTypeBuilder<Skill> builder)
            {
                 builder.ToTable("Skill");

            //Primary key
            builder.HasKey(x => x.Id);

            builder.Property(x => x.Id)
            .ValueGeneratedOnAdd()
            .UseIdentityColumn();

            builder.Property(x => x.SkillName)
            .IsRequired() //Not null
            .HasColumnName("SkillName")
            .HasColumnType("NVARCHAR")
            .HasMaxLength(80);

            builder.Property(x => x.YourSkillScore)
            .IsRequired() //Not null
            .HasColumnName("YourSkillScore")
            .HasColumnType("NVARCHAR")
            .HasMaxLength(80);
        }}}

**Reply by RochaRaphael** [reply]
Quando faço um objeto skill:

       var skill1 = new Skill
                {
                    SkillName = "Java",
                    YourSkillScore = 8
                };

E depois tento adicionar/relacionar esse objeto ao contact:

    var contato1 = new Contact
                {
                    Name = "Raquel",
                    Email = "raquel@outlook.com",
                    Age = 19,
                    Cellphone = "123453339",
                    Skills = skill1;
                };

Ele da erro no skill1 e aparece a seguinte mensagem:
Não é possível converter implicitamente tipo "NetworkingList.Models.Skill" em "System.Collections.Generic.IList<NetworkingList.Models.Skill>". Existe uma conversão explícita

Não sei se essa é a forma correta de se fazer, mas procurando na internet eu não consegui descobri uma saída para fazer usando o FluentMap

**Reply by andrebaltieri** [reply]
Bom dia, @RochaRaphael, eu erro diz que você está tentando atribuir um (Um único objeto) a uma propriedade que espera uma lista. No caso, faz sentido pois pelo que vi no seu mapeamento, você quer uma lista de skills, mas no contato passa apenas um objeto skill.

Não seria assim por exemplo:
```csharp
var contato1 = new Contact
            {
                Name = "Raquel",
                Email = "raquel@outlook.com",
                Age = 19,
                Cellphone = "123453339",
                Skills = new List<Skill> { skill1 };
            };

```

**Reply by RochaRaphael** [reply]
Deu certo, professor! Muito obrigado.
